src/effects/VST3/VST3Effect.cpp

```diff
@@ -339,32 +339,129 @@ bool VST3Effect::SupportsAutomation() const
    return false;
 }
 
-bool VST3Effect::SaveSettings(
-   const EffectSettings &, CommandParameters & parms) const
+struct VST3Effect::ParameterInfo
 {
-   if(mEditController == nullptr)
+   Steinberg::Vst::ParameterInfo mVstParamInfo;
+};
+
+
+bool VST3Effect::ForEachParameter(ParameterVisitor visitor) const
+{
+   if (mEditController == nullptr)
       return false;
 
-   using namespace Steinberg;
-   
-   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   for (int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
    {
-      Vst::ParameterInfo parameterInfo { };
-      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      ParameterInfo pi;
+
+      Steinberg::Vst::ParameterInfo vstParameterInfo{ };
+      if (mEditController->getParameterInfo(i, vstParameterInfo) == Steinberg::kResultOk)
       {
-         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+         if (vstParameterInfo.flags & Steinberg::Vst::ParameterInfo::kCanAutomate)
          {
-            parms.Write(
-               VST3Utils::MakeAutomationParameterKey(parameterInfo),
-               mEditController->getParamNormalized(parameterInfo.id)
-            );
+            if (!visitor(ParameterInfo{ vstParameterInfo } ))
+               return false;
```

Comment:
This is not analogous to what I did for AudioUnits:  the visitor function can return false, not to cause `ForEachParameter` to fail, but only to cause it to stop visiting other parameters.

In fact `ForEachParameter` for AudioUnits ended up as a void-returning function, and yours might too.